Refactoring common functions under pyodide_build

rth opened the pull request:
First attempt to address https://github.com/iodide-project/pyodide/issues/99

Currently this only moves some of the functions from `tools` under the identically named files in the `pyodide_build` package. 

The package might work when installed in developpement mode,
```
pip install -e .
```
but not in a regular install because then `pyodide_build.common.ROOTDIR` will not be correct. Also added a short `setup.py`.

TODO:
 - [x] better mechanism for handling `ROOTDIR` in installed package
 - [x] Continue to use the local copy from make as requested in https://github.com/iodide-project/pyodide/issues/99#issuecomment-410279832

mdboom commented:
I know this is a WIP, but hopefully you don't mind comments along the way.

From my point of view, the real reason to do this is so that other packages outside of the pyodide source tree can build themselves.  We can install the tools as a Python package, but they will still need access to the special native CPython build in `cpython/build/3.6.4/host`, as well as the patched emsdk toolchain.  For a lot of reasons, maybe that doesn't all belong as a Python package.

Maybe instead we need to borrow how `emsdk` works -- you build it somewhere and then to use it you `source` an environment script (`emsdk_env.sh`).  We could easily make a script that makes the pyodide-build scripts and the pyodide-build library available when you do that.

Totally open to other ideas as well.

rth commented:
Thanks for your feedback @mdboom ! It's an interesting idea! I understand we need to set the variables from `Makefile.envs` among other things, and I agree that  the special native CPython and the patched emsdk toolchain probably don't belong in the python package. 

If we make something similar to `emsdk` workflow, I imagine it could be something along the lines of,
```sh
pyodide_build setup   # downloads and builds emsdk & native Python
source ./pyodide_env.sh  # set the environement variables
pyodide_build install <package_name_or_definition_file>
```

Though whether the environment is set with `source`, using e.g.  `pyodide_build activate` (similarly to `conda activate env`) or implicitly when `pyodide_build install` is called might have the same effect in the end.. I will also depend of how to make this use case possible without adding too much redundant code with what is currently in the makefile.

I'm also not sure how much of this would be done in this PR, I was thinking more about small incremental changes...

Will get back to this toward the end of the week..

rth commented on Makefile:
```diff
@@ -204,6 +204,8 @@ ccache/em++:
 $(CPYTHONLIB): emsdk/emsdk/.complete ccache/emcc ccache/em++
 	make -C $(CPYTHONROOT)
+	python -m pip install -e .
+	"$(HOSTPYTHONROOT)/bin/python3" -m pip install -e .
```
We install the `pyodide_build` package both in the regular python and the host python. It just creates a symlink, so it's not too costly. I'm not sure if this is the right place to put it though.

For the system python, this probably assumes that we are in some kind of virtualenv though and this might fail with insufficient permissions if pip attempts to install it somewhere under `/usr/lib`.  The alternative could be to ask users install it explicitly before running make.

mdboom commented on pyodide_build/common.py:
```diff
@@ -1,9 +1,10 @@
-ROOTDIR = Path(__file__).parent.resolve()
-HOSTPYTHON = ROOTDIR / '..' / 'cpython' / 'build' / '3.7.0' / 'host'
-TARGETPYTHON = ROOTDIR / '..' / 'cpython' / 'installs' / 'python-3.7.0'
+ROOTDIR = Path(__file__).parents[1].resolve()
```
Is this going to work if the package is installed, and then the source tree is moved or removed?

rth commented:
Hmm, yes I haven't dealt with toolchains too much and my experience with regular python packages may be biasing my judgement here. Also you are right, the relative `ROOTDIR` definition needs fixing (I have only installed in `develop` mode so far) and I agree with your comment regarding `make install`

So the generally promise of this refactoring in my mind was not so much to change the way package building works, or have a full `emsdk` pipeline with `pip install pyodide_build`. Also having to install `pyodide_build` in `make` is indeed inelegant and I hope there is a way around that. If I there is a way to not affect the way current `make all` works with this PR I would chose it. I agree that using source to get a working environment might be preferable.

On the other hand, putting different python scripts into a separate component in the form of a Python package (that would be in the long term independent of it's location, and the rest of the folder structure) would help with the following points IMO,
 - currently to test any function we have to manually manipulate `sys.path` to add e.g. `./tools/common.py` there. This would not be considered good practice for a Python project, but again maybe I am biased on it, and because this is a build tool the usual principles wouldn't apply.
 - For instance, having a package would allow to do something like,
   ```
   pip install pyodide_build
   pytest tests/  --build /tmp/build/
   ```
   without needing the rest of the setup.
 - `tests/conftest.py` is currently taking a life of it's own, with a full version of the web server. Documenting to run `python tests/conftest.py` to get the web server locally is IMO a bit strange, and e.g. putting some that code into `pyodide_build/web_server.py` and exposing `pyodide start-web-server` (or some other name) could be a way around it.
 - in the long term, when `emsdk` may no longer need patches, having a standalone component with the Python scripts, that could work with a `build/`   `packages/` in any locations might be helpful useful. I'm not saying that it will replace the general Makefile, but rather be used as one of the components there.
 - also in the future, one could for instance imagine some Docker container with the pre-build emsdk toolchain, and another entry point e.g.
   ```
   pyodide build-in-docker packages/  --out /tmp/build/
   ```
   that would take in package definitions and output the build packages without having to re-compile LLVM. That might improve usability. 

Generally, there are a lot of things that could be done given a modular Python package (not saying that all of them are good), but again maybe because I'm a Python developer, I see Python packages everywhere even when there are none :)

Do you think some of that might be possible/useful while keeping your vision of how the build should work (with sourcing a bash script etc)?

rth commented:
After changing `PYTHONPATH` in `Makefile.envs` it is no longer necessary to install `pyodide_build` in make or elsewhere.

Also you are right, this is still not a real stand alone python package, currently only install in develop mode is supported,
```
pip install -e .
```
while
`pip install .` or `python setup.py install` will explicitly fail at setup time.

Overall this aims to make the corresponding scripts more modular (also with respect to `tests/` and `benchmarks/`),
 - by using a single entry-point `python -m build_script <action_name>` (or alternatively `pyodide <action_name>` if the package is installed)
 - using the Python module import system, instead of manually modifying `sys.path` with relative paths in different scripts or tests. This doesn't impact how  `make all` works.

To make this package fully standalone (if we decide to go that way in the future), one would need indeed to,
 - have a way to pass `ROOTDIR` (currently pointing to `tools/`) in the exposed CLI commands
 - use less frequently `__file__` in the scripts

Meanwhile, this would allow to e.g. get rid of `sys.path` manipulations in `benchmark/benchmark.py` if `FirefoxWrapper`, `ChromeWrapper` definitions are also moved into this package in the future.

Would this be more acceptable?

rth commented:
No worries, I appreciate very much your review even when it's not positive.

I removed `setup.py` and added `./bin/pyodide` as you suggest. Please let me know if you would prefer another name.

Regarding building a package externally and documenting it, could this be done in a subsequent PR, as this one already includes a number of various changes? Currently the main issue is that one can't just `source Makefile.envs` because it uses Makefile definition of variables instead of the bash one, so some parsing of that might be needed (or adapting `Makefile.envs` to be a valid bash file).

rth commented on pyodide_build/pywasmcross.py:
```diff
@@ -221,14 +223,18 @@ def parse_args():
+
+
+if __name__ == '__main__':
+    parser = make_parser(argparse.ArgumentParser())
```
Yes, it's a but tricky, because due to the fact that we gave the main parser, to make it work an `ArgumentParser` object need to be provided in any case. I tried to improve the situation a bit and skip all parsing in this case.

For instance, `tools/gcc -h` will still yield the wrong help, it should be disabled but it appears not to be, but everything else we are directly sending arguments to gcc etc.

rth commented:
Test finally pass (except for one `ConnectionResetError: [Errno 104] Connection reset by peer` error in Chrome  we also see randomly in other PRs).

In the next iterations, it might be simpler to just re-write `bin/pyodide` in Python: here I quite spend some time investigating issues due to the difference between dash (default `/bin/sh` shell on Debian based systems) and bash.

mdboom commented on pyodide_build/buildpkg.py:
```diff
@@ -88,7 +85,7 @@ def compile(path, srcpath, pkg, args):
     try:
         subprocess.run([
             str(Path(args.host) / 'bin' / 'python3'),
-            str(ROOTDIR / 'pywasmcross'),
```
should we make this a call to `bin/pyodide`?

rth replied:
We can't make it a call to `bin/pyodide` in an obvious way (at least I haven't found one), because `bin/pyodide` is a shell script that will use the python from the current environment by default, while here we want to use the host Python.

Using `host/bin/python -m pyodide_build pywasmcross` may be on some level lighter, because at this point `PYTHONPATH` was already set, so when we start this subprocess we an use `pyodide_build` directly, there is not need to re-set it in `bin/pyodide`..

mdboom replied:
That's a good point.  Thanks.

rth commented:
Thanks a lot!

rth commented:
Also note that when re-building with this setup for the first time, one would likely experience errors of the form,
```py
  File "/home/rth/src/pyodide/pyodide_build/pywasmcross.py", line 81, in make_symlinks
    symlink_path.symlink_to(exec_path)
  File "/home/rth/src/pyodide/cpython/build/3.7.0/host/lib/python3.7/pathlib.py", line 1309, in symlink_to
    self._accessor.symlink(target, self, target_is_directory)
  File "/home/rth/src/pyodide/cpython/build/3.7.0/host/lib/python3.7/pathlib.py", line 418, in symlink
    return os.symlink(a, b)
FileExistsError: [Errno 17] File exists: '/home/rth/src/pyodide/pyodide_build/pywasmcross.py' -> '/home/rth/src/pyodide/tools/c++'
```
this is because the symlink exists but is broken due to directory name change, so `symlink_path.exists()` will return False. After a quick search I wasn't able to determine how to actually check that a symlink exists without resolving it.. Anyway since it's a one time thing, the easiest might be to `rm tools/{ar,c++,cc,gcc,ld}` manually and let them be re-created.
